tutorial/lesson_24_async.cpp

```diff
@@ -0,0 +1,198 @@
+        producer1.compute_root().async();
+        producer2.compute_root().async();
+
+        // The high-level structure of the generated code will be:
+        // {
+        //     allocate producer1[...]
+        //     allocate producer2[...]
+        //     thread #1 {
+        //         produce producer1 {
+        //             ...
+        //         }
+        //         signal that producer1 data is ready
+        //     }
+        //     thread #2 {
+        //         produce producer2 {
+        //             ...
+        //         }
+        //         signal that producer2 data is ready
+        //     }
+        //     thread #3 {
+        //         consume producer1 {
+        //             consume producer2 {
+        //                 block until producer1 data is ready
+        //                 block until producer2 data is ready
+        //                 produce consumer {
+        //                     ...                        
+        //                 }
+        //             }
+        //         }
+        //     }
+        // }
+        consumer.realize({128, 128});
+    }
+
+    {
+        // In the previous example, we managed to run two producers in parallel, but `consumer` had
+        // to wait until the data is fully ready. Wouldn't it be great if we could overlap computations
+        // of `producer` and `consumer` too? This computational pattern is known as 'double buffering' and
+        // can be critical for achieving good performance in certain scenarios. The high-level idea is that
+        // producer is allowed to run ahead and do the next chunk of work without waiting while consumer 
+        // is processing the current chunk. The obvious drawback of this method is that it requires twice
+        // as much memory for `producer`. 
+        Func producer("producer"), consumer("consumer");
+
+        producer(x, y, c) = (x + y) * (c + 1);
+        consumer(x, y, c) = producer(x - 1, y - 1, c) + producer(x, y, c) + producer(x + 1, y + 1, c);
+
+        consumer.compute_root();
+
+        // In this example the planes are processed separately, so producer can run ahead 
+        // and start producing plane `c + 1`, while `consumer` consumes already produced plane `c`.
+        // One way to express it with Halide schedule is very similar to how sliding window schedules
+        // are expressed (see lesson_8 for details). There are indeed a lot of commonalities between two
+        // because both of them are relying on a circular buffer as underlying data structure.
+        producer.compute_root()
+            .async()
+            .compute_at(consumer, c)
+            .store_at(consumer, Var::outermost())
+            // Explicit fold_storage is required here, because otherwise Halide will infer that only
+            // one plane of `producer` is necessary for `consumer`, but for the purposes of this
+            // example we want at least 2.
+            .fold_storage(c, 2);
+
```

Comment:
A "high-level structure of the generated code" would be welcome.

Reply:
Added a code structure with detailed comments.